compiler/rustc_monomorphize/src/partitioning.rs

```diff
@@ -404,18 +405,53 @@ fn merge_codegen_units<'tcx>(
                 }
             }
         }
+
+        // A sorted order here ensures what follows can be deterministic.
+        codegen_units.sort_by(|a, b| a.name().as_str().cmp(b.name().as_str()));
     } else {
-        // If we are compiling non-incrementally we just generate simple CGU
-        // names containing an index.
+        // When compiling non-incrementally, we rename the CGUS so they have
+        // identical names except for the numeric suffix, something like
+        // `regex.f10ba03eb5ec7975-cgu.N`, where `N` varies.
+        //
+        // It is useful for debugging and profiling purposes if the resulting
+        // CGUs are sorted by name *and* reverse sorted by size. (CGU 0 is the
+        // biggest, CGU 1 is the second biggest, etc.)
+        //
+        // So first we reverse sort by size. Then we generate the names with
+        // zero-padded suffixes, which means they are automatically sorted by
+        // names. The numeric suffix width depends on the number of CGUS:
+        // - 0..=9     CGUS: `0`, `1`, `2`, ...
+        // - 10..=99   CGUS: `00`, `01`, `02`, ...
+        // - 100..=999 CGUS: `000`, `001`, `002`, ...
+        // - And so on.
```

Comment:
Would it make sense to use hex numbers? That is slightly shorter and has a more efficient way to calculate the length.

Reply:
In practice the maximum number of CGUs is almost always 16, so two digits is the most common number. Hex digits would be shorter, but I think keeping it decimal is simpler and one extra digit for this internal name is not a problem.